### PR TITLE
Fix/job manager logger

### LINKED ISSUE #47389: Core: File descriptor of ray job does not get closed
### What happened + What you expected to happen

File descriptor does not get closed in the job manager process on this line when long running job cluster keep receiving job. https://github.com/ray-project/ray/blob/e164d87a49f8b5c7bc5778859873d75845705b47/python/ray/dashboard/modules/job/job_manager.py#L91 leading to high cpu usage 

[
Screenshot 2024-08-28 at 8 08 38 PM
](url)

### Versions / Dependencies

Ray 2.34
Python 3.11

### Reproduction script

Ray jobs on a cluster with some fixed interval (e.g: 30 seconds)

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by ujjawal-khare-27**
## Related issue number

Closes https://github.com/ray-project/ray/issues/47389

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
@GeneDer could you review this PR?

File: python/__init__.py
Comment (GeneDer) [inline python/__init__.py:1]: Is this change needed in this PR?

File: python/ray/dashboard/modules/job/utils.py [hunk 1]
```diff
@@ -302,3 +301,9 @@ async def find_jobs_by_job_ids(
             for job_info, submission_id in zip(job_infos, job_submission_ids)
         },
     }
+
+
+def close_logger_file_descriptor(logger_instance):
+    for handler in logger_instance.handlers:
+        handler.close()
+        logger_instance.removeHandler(handler)
```
Comment (GeneDer) [inline python/ray/dashboard/modules/job/utils.py:309]: I'm fine with this approach, but note that this will cause all future use of this logger becoming no-ops. Based on this [post](https://stackoverflow.com/a/15474586) It seems like just call `handler.close()` is enough to close out the file descriptor. Maybe we can refactor a `close_fd_after_logs()` and that calls `.info` then close the handler so the logger can be used for future?

**Review by GeneDer — COMMENTED** [review]
Would also suggest to add a test case to ensure fd are closed with this approach. 

Also note, I believe @rkooo567 is also working on solving this issue at https://github.com/ray-project/ray/pull/47782 maybe can chat with him on his progress and findings.

File: python/__init__.py
Comment (ujjawal-khare-27) [reply python/__init__.py:1]: Removed

File: python/ray/dashboard/modules/job/utils.py (hunk 1 above)
Comment (ujjawal-khare-27) [reply python/ray/dashboard/modules/job/utils.py]: Hey @GeneDer. I have made following changes to the PR. 

- I have moved to closing of logger to finally block so that info/error can be logged in between if required.
- Moved logger to global context of submit_job function instead of try block so that same logger can be used in exception as well.
- Warning in exception block was being routed to wrong logger, moved it to job logger for easy debugging.
- Added a unit test for the same. Also handler.close is not enough we have to remove handler as well. Tested this as well.


Please let me know your views :)

Comment (ujjawal-khare-27) [reply python/ray/dashboard/modules/job/utils.py]: One problem that I can see  with  `close_fd_after_logs() and that calls .info` is that we have to offer warn, error etc as well. Then we would be making a custom wrapper over logger to provide all functionalities.

File: python/ray/dashboard/utils.py [hunk 2]
```diff
@@ -692,3 +692,9 @@ def compose_state_message(
         else:
             state_message = death_reason_message
     return state_message
+
+
+def close_logger_file_descriptor(logger_instance):
+    for handler in logger_instance.handlers:
+        handler.close()
+        logger_instance.removeHandler(handler)
```
Comment (GeneDer) [inline python/ray/dashboard/utils.py:700]: I think the issue is still we are removing the handlers here. This will cause any log statement after this is called to a logger becoming a no-op

File: python/ray/dashboard/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    job_driver_log_path = "/tmp/ray.log"
+    job_driver_handler = logging.FileHandler(job_driver_log_path)
+    job_driver_formatter = logging.Formatter(logger_format)
+    job_driver_handler.setFormatter(job_driver_formatter)
+    logger.addHandler(job_driver_handler)
+
+    assert len(logger.handlers) == 1
+    close_logger_file_descriptor(logger)
+    assert len(logger.handlers) == 0
```
Comment (GeneDer) [inline python/ray/dashboard/tests/test_utils.py:21]: I don't think we want to remove and check the number of handlers for this. If the behavior is not to have fd open, then that's what we should be checking in the test

File: python/ray/dashboard/modules/job/utils.py [hunk 4]
```diff
@@ -27,7 +27,6 @@
     Request = None
     Response = None
 
-
```
Comment (GeneDer) [inline python/ray/dashboard/modules/job/utils.py:30]: Let's not include unwanted changes.

**Review by GeneDer — COMMENTED** [review]
Few comments, but please also rebase to master, there are currently 852 commits in this PR and a lot of them are unrelated 😅

File: python/ray/dashboard/tests/test_utils.py (hunk 3 above)
Comment (ujjawal-khare-27) [reply python/ray/dashboard/tests/test_utils.py]: Refactored the test for the same

File: python/ray/dashboard/modules/job/utils.py (hunk 4 above)
Comment (ujjawal-khare-27) [reply python/ray/dashboard/modules/job/utils.py]: Reverted

**Comment by ujjawal-khare-27** [comment]
> Few comments, but please also rebase to master, there are currently 852 commits in this PR and a lot of them are unrelated 😅

Tried rebasing multiple times but still additional commits is not going away. Can you please help me in this?

Screenshot 2024-10-16 at 1 13 06 AM

**Comment by GeneDer** [comment]
> > Few comments, but please also rebase to master, there are currently 852 commits in this PR and a lot of them are unrelated 😅
> 
> Tried rebasing multiple times but still additional commits is not going away. Can you please help me in this?
> 
> Screenshot 2024-10-16 at 1 13 06 AM

Maybe you can first pull the latest master, checkout a new branch, then cherry pick your commits from this PR? Or just create new PR with a commit with the changes you have currently, probably easier this way.

**Comment by ujjawal-khare-27** [comment]
@GeneDer opened new PR for cleaner commit history.
Link: https://github.com/ray-project/ray/pull/48029
